### ISSUE TITLE
Type conversion for Hijacker in custom plugins

### ISSUE THREAD

**Issue by fdevillard** [traefik/traefik#8682]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.containo.us) and didn't find any.

### What did you do?

I've implemented a custom middleware that simply wraps a `http.ResponseWriter`. It's main purpose is to investigate how we can handle websockets using middlewares.

The code is available [here](https://github.com/fdevillard/plugin). I've written a test which _validates_ that the `ResponseWriter` can indeed be hijacked (or, at least tries to validate it 😅). The test passes.

When used, I expect no changes in the requests/responses.

### What did you see instead?

The middleware panics with the following logs:

```
{"level":"error","module":"github.com/fdevillard/plugin","msg":"plugins-local/src/github.com/fdevillard/plugin/main.go:21:8: panic","plugin":"plugin-audit","time":"2022-01-06T15:42:54Z"}
{"level":"error","middlewareName":"traefik-internal-recovery","middlewareType":"Recovery","msg":"Recovered from panic in HTTP handler [172.20.0.27:36380 - /graphql]: interface conversion: stdlib._net_http_ResponseWriter is not http.Hijacker: missing method Hijack","time":"2022-01-06T15:42:54Z"}
```

I don't get why it fails to detect the `Hijack` function 🤔 Is it related to Yaegi?

Furthermore, from a frontend perspective, the websocket requests fail.

### What version of Traefik are you using?

Version:      2.5.6
Codename:     brie
Go version:   go1.17.5
Built:        2021-12-22T16:30:52Z
OS/Arch:      linux/amd64

### What is your environment & configuration?

Traefik runs on a kubernetes cluster and is setup with CRDs.

The plugin is enabled using:
```
 --experimental.localPlugins.audit.moduleName=github.com/fdevillard/plugin                                                                                                                                            
```

**Comment by kevinat** [comment]
Hi there,

Is there any further progress on this issue?

**Comment by colearendt** [comment]
@mpl Yes, you're right. I think what you're seeing is ignorance slowly being rolled back 😅 

It wasn't clear to me where the `_net` messages were coming from, so recognizing it was syntax from yaegi was helpful. I'm suspicious that the ResponseWriter is being type asserted to a `ResponseWriter` and is losing the `Hijacker` interface when getting passed either to or from `yaegi`. However, the `ResponseWriter` is a bit tricky to track since it is an argument passed to the Handler's `ServeHTTP` method (the Handler is [definitely type asserted here](https://github.com/traefik/traefik/blob/3c1d5e0393e91ce977ce29d1176c1303d7e6dbf7/pkg/plugins/middlewares.go#L72))

**Comment by colearendt** [comment]
Actually, would this do it? This is where it looks like the type assertion to the `arg` type (in this case a `ResponseWriter`) happens. It may be on passing it back to the main thread though...? It's not clear to me how the "next" call gets back out of Yaegi

https://github.com/traefik/yaegi/blob/fbee2baf9ddfad88f1d9b8c32fc3ca492e5c1ea5/interp/run.go#L1004-L1023

More thorough writeup below:

My current understanding is:
- At the start of the server, all of the plugins are built / compiled / have their `.New` functions called
- As a part of this process, there is an `alice` chain of middlewares built that call one another at the end of their `ServeHTTP` method
- As an aside, there is an interface here that represents what is "needed" for this type of workflow
https://github.com/traefik/traefik/blob/c51e59059195257adba402ece7583083e91bd350/pkg/middlewares/stateful.go#L5-L12
    - Example usage here:
https://github.com/traefik/traefik/blob/c51e59059195257adba402ece7583083e91bd350/pkg/middlewares/customerrors/custom_errors.go#L22-L27
- When a request comes in, the server's `ServeHTTP` method (depending on the entrypoints and routers) ends up calling the alice chain
https://github.com/traefik/traefik/blob/c51e59059195257adba402ece7583083e91bd350/pkg/server/router/router.go#L129-L136
- If the chain has only "proper middlewares," things work well. They are called something like this:
https://github.com/traefik/traefik/blob/c51e59059195257adba402ece7583083e91bd350/pkg/server/middleware/middlewares.go#L191-L199
- If the chain has "plugins," they go through a different initialization process:
https://github.com/traefik/traefik/blob/c51e59059195257adba402ece7583083e91bd350/pkg/server/middleware/middlewares.go#L353-L361
- In particular, the `plug` handler represents the output of the arguments getting passed to Yaegi for the Plugin's "New" function to be called
https://github.com/traefik/traefik/blob/c51e59059195257adba402ece7583083e91bd350/pkg/plugins/middlewares.go#L62-L77
- When calling `run.go` inside of Yaegi, we see the args type cast into the "defined" type (i.e. `ResponseWriter`)
https://github.com/traefik/yaegi/blob/fbee2baf9ddfad88f1d9b8c32fc3ca492e5c1ea5/interp/run.go#L1004-L1023
- This stream can also be seen inside of the stack trace above. I have somehow convinced myself that this is an orthogonal stack to the `New` function call I reference above:
```
reflect.Value.call({0x2f9cd60, 0xc00118c050, 0xc00129c4b0}, {0x355d780, 0x4}, {0xc0014fc5a0, 0x2, 0x7f727ab82068})
        /usr/local/golang/1.10.8/go/src/reflect/value.go:543 +0x814
reflect.Value.Call({0x2f9cd60, 0xc00118c050, 0x4517f2}, {0xc0014fc5a0, 0x2, 0x2})
        /usr/local/golang/1.10.8/go/src/reflect/value.go:339 +0xc5
github.com/traefik/yaegi/interp.callBin.func1({0x2f9cd60, 0xc00118c050, 0x4ac5c6}, {0xc0014fc5a0, 0xc00118c040, 0x4ae4dc})
        /home/semaphore/go/pkg/mod/github.com/traefik/yaegi@v0.11.2/interp/run.go:1472 +0x28
github.com/traefik/yaegi/interp.callBin.func10(0xc00136e420)
        /home/semaphore/go/pkg/mod/github.com/traefik/yaegi@v0.11.2/interp/run.go:1646 +0x158
github.com/traefik/yaegi/interp.runCfg(0xc00098b8c0, 0xc00136e420, 0x39, 0x3453a60)
        /home/semaphore/go/pkg/mod/github.com/traefik/yaegi@v0.11.2/interp/run.go:201 +0x2ac
github.com/traefik/yaegi/interp.genFunctionWrapper.func2.1({0xc00120fe30, 0x2, 0x4})
        /home/semaphore/go/pkg/mod/github.com/traefik/yaegi@v0.11.2/interp/run.go:1023 +0x4a5
github.com/traefik/yaegi/stdlib._net_http_Handler.ServeHTTP(...)
        /home/semaphore/go/pkg/mod/github.com/traefik/yaegi@v0.11.2/stdlib/go1_17_net_http.go:288
```

This seems like the expected behavior from Yaegi. However, it does break our chain because the `Hijacker` interface would be lost on the type assertion to `ResponseWriter`. This is really difficult though, because `Handler` is not what we need to fix - we must fix `ResponseWriter`, which is passed when `ServeHTTP` is called by the "previous middleware in the chain." i.e.

https://github.com/traefik/traefik/blob/c51e59059195257adba402ece7583083e91bd350/pkg/middlewares/metrics/metrics.go#L121

Maybe having the plugin define the type of its `ServeHTTP` method to be an interface that extends `ResponseWriter` (i.e. `Stateful`) would solve the problem / test this theory? However, we will need such an interface inside of Yaegi - could I define it in the plugin, or would it need to be inside of Yaegi's source? If does need to be in Yaegi's source, how do we reference it in the plugin so that Yaegi picks it up properly? It seems (from docs) that Yaegi does not support modules, so does that mean that the base library needs to be patched with such an interface?

Does this sound like a reasonable interpretation of the lay of the land?

**Comment by mpl** [comment]
It does seem like it is related to yaegi.
I opened the corresponding issue with a simpler reproduction: https://github.com/traefik/yaegi/issues/1425

**Comment by colearendt** [comment]
So much simpler 😅 That’s impressive, thanks!

**Comment by traefiker** [comment]
Closed by #9327.
